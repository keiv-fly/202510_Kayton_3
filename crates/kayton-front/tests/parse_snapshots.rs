use insta::assert_snapshot;
use kayton_front::tests_support::parse_str;

#[test]
fn parse_simple_function() {
    let source = r#"fn add(a, b):
    let total = a + b
    total
"#;
    let output = parse_str("add.ktn", source);
    assert_snapshot!(
        "simple_function",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_if_else_suite() {
    let source = r#"fn classify(x):
    if x > 0:
        "positive"
    elif x < 0:
        "negative"
    else:
        "zero"
"#;
    let output = parse_str("classify.ktn", source);
    assert_snapshot!(
        "if_else_suite",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_inline_if_expression() {
    let source = r#"fn absolute(x):
    let sign = if x < 0: -1 else: 1
    if sign == -1:
        -x
    else:
        x
"#;
    let output = parse_str("absolute.ktn", source);
    assert_snapshot!(
        "inline_if",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_while_loop() {
    let source = r#"fn countdown(n):
    let current = n
    while current > 0:
        let _ = current
        let current = current - 1
        ()
    ()
"#;
    let output = parse_str("countdown.ktn", source);
    assert_snapshot!(
        "while_loop",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}
