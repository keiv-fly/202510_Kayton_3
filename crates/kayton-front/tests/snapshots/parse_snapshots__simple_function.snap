---
source: crates/kayton-front/tests/parse_snapshots.rs
assertion_line: 11
expression: "format!(\"{:#?}\\n---\\n{:#?}\", output.module, output.diagnostics)"
---
HirModule {
    id: HirId(
        1,
    ),
    items: [
        Function(
            HirFunction {
                id: HirId(
                    2,
                ),
                name: Symbol(
                    0,
                ),
                params: [
                    HirParam {
                        id: HirId(
                            3,
                        ),
                        name: Symbol(
                            1,
                        ),
                        span: Span {
                            source: SourceId(
                                1,
                            ),
                            start: 7,
                            end: 8,
                        },
                    },
                    HirParam {
                        id: HirId(
                            4,
                        ),
                        name: Symbol(
                            2,
                        ),
                        span: Span {
                            source: SourceId(
                                1,
                            ),
                            start: 10,
                            end: 11,
                        },
                    },
                ],
                body: HirBlock {
                    id: HirId(
                        5,
                    ),
                    statements: [
                        Let(
                            HirLetBinding {
                                id: HirId(
                                    6,
                                ),
                                name: Symbol(
                                    3,
                                ),
                                value: Binary(
                                    HirBinary {
                                        id: HirId(
                                            7,
                                        ),
                                        op: Add,
                                        lhs: Name(
                                            HirNameRef {
                                                id: HirId(
                                                    8,
                                                ),
                                                name: Symbol(
                                                    1,
                                                ),
                                                span: Span {
                                                    source: SourceId(
                                                        1,
                                                    ),
                                                    start: 30,
                                                    end: 31,
                                                },
                                            },
                                        ),
                                        rhs: Name(
                                            HirNameRef {
                                                id: HirId(
                                                    9,
                                                ),
                                                name: Symbol(
                                                    2,
                                                ),
                                                span: Span {
                                                    source: SourceId(
                                                        1,
                                                    ),
                                                    start: 34,
                                                    end: 35,
                                                },
                                            },
                                        ),
                                        span: Span {
                                            source: SourceId(
                                                1,
                                            ),
                                            start: 30,
                                            end: 35,
                                        },
                                    },
                                ),
                                span: Span {
                                    source: SourceId(
                                        1,
                                    ),
                                    start: 18,
                                    end: 35,
                                },
                            },
                        ),
                    ],
                    tail: Some(
                        Name(
                            HirNameRef {
                                id: HirId(
                                    10,
                                ),
                                name: Symbol(
                                    3,
                                ),
                                span: Span {
                                    source: SourceId(
                                        1,
                                    ),
                                    start: 40,
                                    end: 45,
                                },
                            },
                        ),
                    ),
                    span: Span {
                        source: SourceId(
                            1,
                        ),
                        start: 12,
                        end: 46,
                    },
                },
                span: Span {
                    source: SourceId(
                        1,
                    ),
                    start: 0,
                    end: 46,
                },
            },
        ),
    ],
    interner: SymbolInterner {
        strings: {
            "add",
            "a",
            "b",
            "total",
        },
    },
}
---
[]
